File: Leet322_CoinChange.cpp
/*

Leet322_CoinChange

Dynamic programming

*/

//Brute Force
//using Recursive call


#include <iostream>
#include <string>
#include <vector>
#include <cmath>

using namespace std;


int coinChange(vector<int>& coins, int amount){
    
    int mem[amount+1];
    mem[0]=0;
    
    sort(coins.begin(), coins.end());
    
    for(int currentAmount =1; currentAmount<=amount; currentAmount++){
        mem[currentAmount] = 1000000000;
        for(int i=0; i< coins.size(); ++i){
            if(currentAmount>= coins[i] && mem[currentAmount-coins[i]]!=-1){
                mem[currentAmount] = min( mem[currentAmount-coins[i]]+1, mem[currentAmount]);
            }
        }
        if(mem[currentAmount] == 1000000000)
            mem[currentAmount] =-1;
    }
    return mem[amount];
}


int main(){
    vector<int> eg = {2};
    int amount =3;
    
    cout<<coinChange(eg, amount);
    
    
}

File: Leet322_CoinChange.java
import java.util.ArrayList;

public class Leet322_CoinChange {
    public int coinChange(int[] coins, int amount) {
        int row = amount+1, col = coins.length;
        if (row < 1 || amount < 1) return -1;

        int[][] dp = new int[row][col]; // java initialize with Zero

        for (int j = 1; j<row; j++){
            int curMin = Integer.MAX_VALUE;
            for (int i = 0; i< col; i++){
                int curAmount = coins[i];
                if(j >= curAmount && dp[j-curAmount][i] != Integer.MAX_VALUE ){
                    dp[j][i] = Math.min(dp[j-curAmount][i]+1, curMin);
                    curMin = Math.min(curMin, dp[j][i]);
                }else{
                    dp[j][i] = curMin;
                }
            }
        }
        return dp[row-1][col-1] == Integer.MAX_VALUE ? -1 : dp[row-1][col-1] ;
    }
}
